### PR TITLE
HID: Fixes rumble only in one controller

### PR DESCRIPTION
**Pull request by german77**
Fixed regression by one of the last stages of #4291.
Handles multiple vibrations in one command fixing vibration on 1-2 switch.
If a controller is mapped as a left joycon it will switch the button to DDown.

### CODE REVIEW

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 1]
```diff
@@ -673,25 +673,41 @@ void Controller_NPad::VibrateController(const std::vector<u32>& controllers,
         return;
     }
     bool success = true;
-    for (std::size_t i = 0; i < controllers.size(); ++i) {
-        if (!connected_controllers[i].is_connected) {
-            continue;
+    if (vibrations.size() == 1) {
+        for (std::size_t i = 0; i < controllers.size(); ++i) {
+            std::size_t controller = (controllers[i] >> 8) & 0x7;
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/npad.cpp:678]:
```suggestion
            const std::size_t controller = (controllers[i] >> 8) & 0x7;
```

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 2]
```diff
@@ -673,25 +673,41 @@ void Controller_NPad::VibrateController(const std::vector<u32>& controllers,
...
-        const auto& button_state = buttons[i];
-        if (button_state[A - BUTTON_HID_BEGIN]) {
-            if (button_state[A - BUTTON_HID_BEGIN]->SetRumblePlay(
-                    vibrations[0].amp_high, vibrations[0].amp_low, vibrations[0].freq_high,
-                    vibrations[0].freq_low)) {
-                success = false;
-            }
+    } else {
+        for (std::size_t i = 0; i < vibrations.size(); ++i) {
+            std::size_t controller = i >> 1;
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/npad.cpp:683]:
```suggestion
            const std::size_t controller = i / 2;
```

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 3]
```diff
@@ -673,25 +673,41 @@ void Controller_NPad::VibrateController(const std::vector<u32>& controllers,
...
 }
 
+bool Controller_NPad::VibratePhysicalController(std::size_t controller, Vibration vibration) {
+    if (!connected_controllers[controller].is_connected) {
+        return true;
+    }
+    using namespace Settings::NativeButton;
+    const auto& button_state = buttons[controller];
+    const auto controller_type = connected_controllers[controller].type;
+    auto use_button = A;
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/npad.cpp:699]:
```cpp
const auto use_button = controller_type == NPadControllerType::JoyLeft ? DDown : A;
```

A switch with only one case in it isn't very useful.

File: src/core/hle/service/hid/controllers/npad.h [hunk 4]
```diff
@@ -128,6 +128,9 @@ class Controller_NPad final : public ControllerBase {
     void VibrateController(const std::vector<u32>& controllers,
                            const std::vector<Vibration>& vibrations);
 
+    // Called from VibrateController and Vibrates a specific controller
```
Comment (lioncash) [inline src/core/hle/service/hid/controllers/npad.h:131]: Where it's called from isn't particularly important. This part of the comment can also become outdated in the future if code gets moved around.

File: src/core/hle/service/hid/controllers/npad.h [hunk 5]
```diff
@@ -128,6 +128,9 @@ class Controller_NPad final : public ControllerBase {
     void VibrateController(const std::vector<u32>& controllers,
                            const std::vector<Vibration>& vibrations);
 
+    // Vibrates a specific controller
+    bool VibratePhysicalController(std::size_t controllers, Vibration vibrations);
```
Comment (v1993) [inline src/core/hle/service/hid/controllers/npad.h:132]: `controllers` -> `controller`

**Review by v1993 — CHANGES_REQUESTED** [review]

**Comment by v1993** [comment]
What's the reason behind closing this?

**Comment by german77** [comment]
It will be superseded by a proper fix. Since this will not be merged or added to EA there is no point on keeping it open

**Pull request closed without merging**